libdnf/base/transaction.cpp

```diff
@@ -0,0 +1,299 @@
+        auto obs = goal.list_obsoleted_by_package(id);
+        if (obs.empty()) {
+            throw RuntimeError("No obsoletes for reinstall");
+        }
+        rpm::Package new_package(sack, rpm::PackageId(id));
+        auto reason = new_package.get_reason();
+        TransactionPackageItem item(new_package, TransactionPackageItem::Action::UPGRADE, reason);
+        item.replace.emplace(rpm::Package(sack, rpm::PackageId(obs[0])));
+        for (int i = 1; i < obs.size(); ++i) {
+            rpm::Package obsoleted(sack, rpm::PackageId(obs[i]));
+            auto obs_reson = obsoleted.get_reason();
+            if (obs_reson > reason) {
+                reason = obs_reson;
+            }
+            item.obsoletes.emplace_back(std::move(obsoleted));
+            obsoletes[obs[i]].push_back(id);
+        }
+        item.set_reason(reason);
+        packages.emplace_back(std::move(item));
+
+        TransactionPackageItem old_item(
+            rpm::Package(sack, rpm::PackageId(obs[0])), TransactionPackageItem::Action::UPGRADED, reason);
+        packages.emplace_back(std::move(old_item));
+    }
+    auto list_removes = goal.list_removes();
+    if (!list_removes.empty()) {
+        rpm::PackageQuery remaining_installed(base, rpm::PackageQuery::InitFlags::IGNORE_EXCLUDES);
+        remaining_installed.filter_installed();
+        for (auto index = 0; index < list_removes.size(); ++index) {
+            remaining_installed.p_impl->remove(list_removes[index]);
+        }
+        rpm::PackageSet tmp_set(base->get_rpm_package_sack());
+
+        // https://bugzilla.redhat.com/show_bug.cgi?id=1921063
+        // To keep a reason of installonly pkgs in DB for remove step it requires TSI with reason change
+        for (auto index = 0; index < list_removes.size(); ++index) {
+            Id id = list_removes[index];
+            rpm::Package rm_package(sack, rpm::PackageId(id));
+            tmp_set.add(rm_package);
+            rpm::PackageQuery remaining_na(remaining_installed);
+            remaining_na.filter_name_arch(tmp_set);
+            if (!remaining_na.empty()) {
+                auto  keep_reason = (*remaining_na.begin()).get_reason();
+                TransactionPackageItem keep_reason_item(*remaining_na.begin(), TransactionPackageItem::Action::REASON_CHANGE, keep_reason);
+                packages.emplace_back(std::move(keep_reason_item));
+            }
+            tmp_set.clear();
+            auto reason = goal.get_reason(id);
+            TransactionPackageItem item(rm_package, TransactionPackageItem::Action::REMOVE, reason);
+            packages.emplace_back(std::move(item));
+        }
+    }
+
+    // Add obsoleted packages
+    for (auto & item : obsoletes) {
+        rpm::Package obsoleted(sack, rpm::PackageId(item.first));
+        auto reason = goal.get_reason(item.first);
+        TransactionPackageItem tsi(obsoleted, TransactionPackageItem::Action::OBSOLETED, reason);
+        for (auto id : item.second) {
```

Comment:
Tip: Using structured bindings is more readable than `item.first`, `item.second`.
```
for (const auto & [obsoleted_id, obsoleted_by_ids] : obsoletes) {
        rpm::Package obsoleted(sack, rpm::PackageId(obsoleted_id));
        auto reason = goal.get_reason(obsoleted_id);
        TransactionPackageItem tsi(obsoleted, TransactionPackageItem::Action::OBSOLETED, reason);
        for (auto id : obsoleted_by_ids) {
```